src/ripple/app/ledger/impl/InboundLedger.cpp

```diff
@@ -972,39 +979,47 @@ InboundLedger::takeAsNode(
         return true;
     }
 
-    auto nodeIDit = nodeIDs.cbegin();
-    auto nodeDatait = data.begin();
-    AccountStateSF filter(
-        mLedger->stateMap().family().db(), app_.getLedgerMaster());
-
-    while (nodeIDit != nodeIDs.cend())
+    try
     {
-        if (nodeIDit->isRoot())
+        auto nodeIDit = nodeIDs.cbegin();
+        auto nodeDatait = data.begin();
+        AccountStateSF filter(
+            mLedger->stateMap().family().db(), app_.getLedgerMaster());
+
+        while (nodeIDit != nodeIDs.cend())
         {
-            san += mLedger->stateMap().addRootNode(
-                SHAMapHash{mLedger->info().accountHash},
-                makeSlice(*nodeDatait),
-                snfWIRE,
-                &filter);
-            if (!san.isGood())
+            if (nodeIDit->isRoot())
             {
-                JLOG(m_journal.warn()) << "Bad ledger header";
-                return false;
+                san += mLedger->stateMap().addRootNode(
+                    SHAMapHash{mLedger->info().accountHash},
+                    makeSlice(*nodeDatait),
+                    &filter);
+                if (!san.isGood())
+                {
+                    JLOG(m_journal.warn()) << "Unable to add AS root node";
+                    return false;
+                }
             }
-        }
-        else
-        {
-            san += mLedger->stateMap().addKnownNode(
-                *nodeIDit, makeSlice(*nodeDatait), &filter);
-            if (!san.isGood())
+            else
             {
-                JLOG(m_journal.warn()) << "Unable to add AS node";
-                return false;
+                san += mLedger->stateMap().addKnownNode(
+                    *nodeIDit, makeSlice(*nodeDatait), &filter);
+                if (!san.isGood())
+                {
+                    JLOG(m_journal.warn()) << "Unable to add AS node";
+                    return false;
+                }
```

Comment:
The loops in `TakeTxNode` and `TakeAsNode` can be improved....

```
        auto& stateMap{mLedger->stateMap()};
        AccountStateSF filter(stateMap.family().db(), app_.getLedgerMaster());
        for (auto& [nodeID, data] : nodes)
        {
            if (nodeID.isRoot())
                san += stateMap.addRootNode(
                    SHAMapHash{mLedger->info().accountHash},
                    makeSlice(data),
                    &filter);
            else
                san += stateMap.addKnownNode(nodeID, makeSlice(data), &filter);

            if (!san.isGood())
            {
                JLOG(m_journal.warn()) << "Unable to add AS node";
                return false;
            }
        }
```

You can cherry-pick my changes here: https://github.com/miguelportilla/rippled/commits/catchy-plus-plus